### ISSUE TITLE
why leave UNKNOWN_TOKEN in the encoder input and decoder target?

### ISSUE THREAD

**Issue by bhomass** [rohithreddy024/Text-Summarizer-Pytorch#15]
When I tested out the model trained on OpenNMT Gigaword on some randomly chosen sentences, I sometimes see <unk> in the response. This got me looking into how OOV words are handled in the code. 

It appears OOV words do make into the extended Vocab in the trail section, however, they are left as UNKNOWN_TOKEN in the encoder input and decoder target. I am having a hard time rationalize that design choice. Obviously, UNKNOWN_TOKEN won't have any meaningful contextual semantics to offer the encoder-decoder process. And having UNKNOWN_TOKEN in the target will make the model strive to generate <unk> during inference based on the MLE loss. The RL loss prob mitigate this to a degree. As the result shows, <unk> does show up fairly often during the experiment.

I am wondering, is this the common way of handling OOV? If you at least allow the target list to point to the OOV word in the extended Vocab list, wouldn't that force the model to generate a summary words out of the extended Vocab instead of spitting out an <unk>?

**Comment by rohithreddy024** [comment]
You mentioned that OOV words are left as UNKNOWN_TOKEN in encoder input and decoder target. For encoder input, OOVs must be replaced with [UNK] tokens as we don't have word embeddings (inputs to encoder LSTM) for OOV words whereas you have word embedding for [UNK]. Same is the case for decoder input.
For decoder target, words can be from vocab, OOV (words not in the vocab but present in the article), UNK (words not in vocab and not in OOV). That's why you see UNK tokens in decoder target. Refer [L46](https://github.com/rohithreddy024/Text-Summarizer-Pytorch/blob/master/data_util/batcher.py#L46) and [L49](https://github.com/rohithreddy024/Text-Summarizer-Pytorch/blob/master/data_util/batcher.py#L49) in `batcher.py` file. But UNK should occur rarely in decoder target and not as frequently as you mentioned. Can you share your input (article) and output (predicted abstract) of the examples you tested.

**Comment by bhomass** [comment]
I follow your argument, but don't quite see why it can't be improved. After all, you are using a pretrained word embedding, like GLOVE, is that right? So why couldn't you convert any article OOV into an embedding on the fly? Since you added the word to the extended OOV vocab, you also have a valid word index to it. I am wondering if you could even apply the word embedding on the fly for target OOV words as well, and extend the vocab even further based on the training abstract. 

Here is an example of extreme use of unk in the output.
"Hello Everyone, this is Bruce here." leads to "<unk> <unk> <unk> <unk> <unk> <unk> <unk>"

**Comment by rohithreddy024** [comment]
The only way you can provide an index to OOV and link it to an embedding is by increasing vocab size & making sure that all OOVs are present in vocab. And increasing vocab size leads to huge memory & computation requirement. 
Regarding your example, I didn't encounter such kind of output before. Check whether your model is properly trained and changes you might have done are bug free. Later first try on gigaword test examples instead of random sentences. If model gives reasonable output on it, then model is trained correctly.

**Comment by bhomass** [comment]
Actually, no. it won't keep growing the vocab if the extended oov only shows up in memory during training, as in [L179](https://github.com/rohithreddy024/Text-Summarizer-Pytorch/blob/master/model.py#L179), where vocab_dist is extended with the article OOV's. I am suggesting, in the encoder input, instead of converting an oov word to UNKNOWN_TOKEN in the original vocab, maybe you can use a idx pointing to the in memory extended vocab_dist.

As for my trained model, when I feed in article "russia 's lower house of parliament was scheduled friday to debate an appeal to the prime minister that challenged the right of u.s.-funded radio liberty to operate in russia following its introduction of broadcasts targeting chechnya", I get a good abstract "russian parliament to debate u.s. radio liberty". That means the model is reasonably trained.

**Comment by rohithreddy024** [comment]
You cannot simply use extended vocab_dict method for encoder input as vocab_dict is directly linked with embedding matrix [L190](https://github.com/rohithreddy024/Text-Summarizer-Pytorch/blob/master/model.py#L190). Modifying vocab_dict (as in extended vocab_dict) also requires modification of embedding matrix while training which is not possible. 
Since the model is trained on Gigaword dataset, the model performs well for inputs which has similar style as that of Gigaword (news articles) and not some random sentences. If you wish the model to give decent ouputs for random sentences, create your own random examples, add them in training data & train the model with it.

**Comment by bhomass** [comment]
ok, I see your point. I was under the impression the embedding is initialized with GLOVE and then kept frozen. That makes it easier to find embedding representation for any new words coming in future articles. Accepting that GLOVE is already general enough, that might be a good direction to try.

**Comment by bhomass** [comment]
Hi, I am still on this point, because I am getting all these [unk] during inference, and the model is just not workable. If I understand your code correctly, the oov's are simply appended at the end of the final_dist as available indices. oov words never have a vector representation during training or inference. That being the case, I don't see how the model will ever learn to point to them. In act, the model is trained to spit out [unk] whenever, they show up in the abstract in the training dataset. So, the abundant [unk]s I am observing is expected behavior for this model. Are we in agreement here?

**Comment by rohithreddy024** [comment]
As I said earlier you can't expect reasonable outputs by training on Gigaword dataset and testing on complete random sentences (which are not similar to news articles). This is true for any machine learning model. Create your own dataset with article & abstract in the style you want (with sufficient examples) and train the model with it.
You don't need OOV vector representation during inference as you just have to sample words from the extended_vocab distribution (which contains OOVs). So the model should output OOVs if they have lot of probability in the distribution.
The model can output [unk]s, but it should happen rarely. If its happening as frequently as you said, then you are testing the model against kind of examples which the model is not trained for.

**Comment by bhomass** [comment]
I have two problems with your explanation. First of all, it is highly desirable to make a trained model cross domain, because something like labeled summarization dataset is just about non-existent, and would be extremely expensive to generate. Basically, that is not a practical solution to suggest to anyone, except maybe Google or Amazon.

Second, my point about the oov implementation is that the model will never assign reasonable probability to accentuate these oov words, because they were never accentuated during training. So the extended_vocab is really just therefore for show. Effectively, they are as good as if they don't exist. 

The point that for a test data in a similar domain will rarely output [unk], even that I find unacceptable, since [unk] is not satisfactory under any circumstances. You are actually better off going to the next highest prob word then to stay with [unk].  For domains which happens to have a common vocabulary size much greater than 50,000, it will kick out [unk] more than just infrequently. It is strictly a matter of luck how much the first 50000 words from the vocabulary list match the actual corpus. 

What I am talking about is going beyond what the status quo performance achieved by the paper, and actually produce a model that can have some real use in the real world. And I am picking your brain how we can achieve that using the current code base as a starting point.
